development/test-log-collector/cmd/app/mainerr.go

```diff
@@ -0,0 +1,259 @@
+		if err != nil {
+			return errors.Wrapf(err, "while getting dispatching config for %s test suite", testName)
+		}
+
+		status, err := extractTestStatus(testName, newestCts)
+		if err != nil {
+			return errors.Wrapf(err, "while extracting test status from ClusterTestSuite for label %s", testName)
+		}
+
+		if status == octopusTypes.TestSucceeded && testConfig.OnlyReportFailure {
+			logf.Infof("skipping report of %s test suite because it has status %s", testName, string(status))
+			continue
+		}
+
+		container, err := getTestContainerName(pod)
+		if err != nil {
+			return errors.Wrapf(err, "while extracting test container name from pod %s in namespace %s", pod.Name, pod.Namespace)
+		}
+		logf.Info(fmt.Sprintf("Extracting logs from container %s from pod %s from namespace %s", container, pod.Name, pod.Namespace))
+		req := clientset.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{
+			Container: container,
+		})
+
+		msg := pkgSlack.Message{
+			Data:        "",
+			PodName:     pod.Name,
+			ChannelName: testConfig.ChannelName,
+			ChannelID:   testConfig.ChannelID,
+			Attributes: pkgSlack.Attributes{
+				Name:             testName,
+				Status:           string(status),
+				ClusterTestSuite: newestCts.Name,
+				CompletionTime:   newestCts.Status.CompletionTime.String(),
+				Platform:         string(platform),
+			},
+		}
+
+		data, err := ConsumeRequest(req)
+		if err != nil {
+			// one of the possible things in test is like a failed init container, in which case we send error as string
+			reqErr := errors.Wrapf(err, "while reading request from container %s in pod %s in namespace %s", container, pod.Name, pod.Namespace)
+			logf.Warnf("couldn't read logs from pod %s, error: %s", pod.Name, reqErr)
+			msg.Data = reqErr.Error()
+		} else {
+			msg.Data = string(data)
+		}
+
+		messages = append(messages, msg)
+	}
+
+	logf.Info("Uploading logs to appropriate slack thread")
+	if err := slackClient.UploadLogFiles(messages, newestCts.Name, newestCts.Status.CompletionTime.String(), string(platform)); err != nil {
+		return errors.Wrap(err, "while uploading files to slack thread")
+	}
+	return nil
+}
+
+func getRestConfigOrDie() *restclient.Config {
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		client, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			panic(errors.Wrapf(err, "while creating restclient based on KUBECONFIG=%s", kubeconfig))
+		}
+		return client
+	}
+
+	client, err := restclient.InClusterConfig()
+	if err != nil {
+		panic(errors.Wrap(err, "while creating in cluster config"))
+	}
+	return client
+}
+
+func getTestContainerName(pod corev1.Pod) (string, error) {
+	names := []string{}
+	for _, cont := range pod.Spec.Containers {
+		if cont.Name != "istio-proxy" {
+			names = append(names, cont.Name)
+		}
+	}
+
+	if len(names) != 1 {
+		return "", fmt.Errorf("found more than 1 non-istio containers in pod %s in namespace %s", pod.Name, pod.Namespace)
```

Comment:
What if some test have to more than 1 container with test logic?

Reply:
Writing tests with yet another sidecar would be pure agony, given all the magic with bash scripts and killing istio-proxy at the end of tests. Even if, then how am I supposed to get the correct one? Tests don't need sidecar other than istio-proxy most of the time, the fact that we have something like 20 tests and none of them has additional sidecar proves this enough for me